src/shogun/mathematics/linalg/LinalgBackendEigen.h

```diff
@@ -99,711 +98,321 @@ class LinalgBackendEigen : public LinalgBackendBase
-	{  \
-		return sum_symmetric_impl(a, no_diag); \
-	}
+	virtual Type sum_symmetric(const linalg::Block<Container<Type>>& a, bool no_diag) const;
 	DEFINE_FOR_ALL_PTYPE(BACKEND_GENERIC_SYMMETRIC_BLOCK_SUM, SGMatrix)
 	#undef BACKEND_GENERIC_SYMMETRIC_BLOCK_SUM
 
 	/** Implementation of @see LinalgBackendBase::colwise_sum */
 	#define BACKEND_GENERIC_COLWISE_SUM(Type, Container) \
-	virtual SGVector<Type> colwise_sum(const Container<Type>& a, bool no_diag) const \
-	{  \
-		return colwise_sum_impl(a, no_diag); \
-	}
+	virtual SGVector<Type> colwise_sum(const Container<Type>& a, bool no_diag) const;
 	DEFINE_FOR_ALL_PTYPE(BACKEND_GENERIC_COLWISE_SUM, SGMatrix)
 	#undef BACKEND_GENERIC_COLWISE_SUM
 
 	/** Implementation of @see LinalgBackendBase::colwise_sum */
 	#define BACKEND_GENERIC_BLOCK_COLWISE_SUM(Type, Container) \
-	virtual SGVector<Type> colwise_sum(const linalg::Block<Container<Type>>& a, bool no_diag) const \
-	{  \
-		return colwise_sum_impl(a, no_diag); \
-	}
+	virtual SGVector<Type> colwise_sum(const linalg::Block<Container<Type>>& a, bool no_diag) const;
 	DEFINE_FOR_ALL_PTYPE(BACKEND_GENERIC_BLOCK_COLWISE_SUM, SGMatrix)
 	#undef BACKEND_GENERIC_BLOCK_COLWISE_SUM
 
 	/** Implementation of @see LinalgBackendBase::rowwise_sum */
 	#define BACKEND_GENERIC_ROWWISE_SUM(Type, Container) \
-	virtual SGVector<Type> rowwise_sum(const Container<Type>& a, bool no_diag) const \
-	{  \
-		return rowwise_sum_impl(a, no_diag); \
-	}
+	virtual SGVector<Type> rowwise_sum(const Container<Type>& a, bool no_diag) const;
 	DEFINE_FOR_ALL_PTYPE(BACKEND_GENERIC_ROWWISE_SUM, SGMatrix)
 	#undef BACKEND_GENERIC_ROWWISE_SUM
 
 	/** Implementation of @see LinalgBackendBase::rowwise_sum */
 	#define BACKEND_GENERIC_BLOCK_ROWWISE_SUM(Type, Container) \
-	virtual SGVector<Type> rowwise_sum(const linalg::Block<Container<Type>>& a, bool no_diag) const \
-	{  \
-		return rowwise_sum_impl(a, no_diag); \
-	}
+	virtual SGVector<Type> rowwise_sum(const linalg::Block<Container<Type>>& a, bool no_diag) const;
 	DEFINE_FOR_ALL_PTYPE(BACKEND_GENERIC_BLOCK_ROWWISE_SUM, SGMatrix)
 	#undef BACKEND_GENERIC_BLOCK_ROWWISE_SUM
 
 	/** Implementation of @see LinalgBackendBase::trace */
 	#define BACKEND_GENERIC_TRACE(Type, Container) \
-	virtual Type trace(const Container<Type>& A) const \
-	{  \
-		return trace_impl(A); \
-	}
+	virtual Type trace(const Container<Type>& A) const;
 	DEFINE_FOR_ALL_PTYPE(BACKEND_GENERIC_TRACE, SGMatrix)
 	#undef BACKEND_GENERIC_TRACE
 
 	/** Implementation of @see LinalgBackendBase::zero */
 	#define BACKEND_GENERIC_ZERO(Type, Container) \
-	virtual void zero(Container<Type>& a) const \
-	{  \
-		zero_impl(a); \
-	}
+	virtual void zero(Container<Type>& a) const;
 	DEFINE_FOR_ALL_PTYPE(BACKEND_GENERIC_ZERO, SGVector)
 	DEFINE_FOR_ALL_PTYPE(BACKEND_GENERIC_ZERO, SGMatrix)
 	#undef BACKEND_GENERIC_ZERO
 
 	#undef DEFINE_FOR_ALL_PTYPE
+	#undef DEFINE_FOR_REAL_PTYPE
+	#undef DEFINE_FOR_NON_INTEGER_PTYPE
+	#undef DEFINE_FOR_NUMERIC_PTYPE
```

Comment:
yep.... indeed. i dont think it's a problem... i mean that the header would only be actually included consciously one would presume :) i.e. this is just to avoid copy-paste errors and actual copy-pasting things around.